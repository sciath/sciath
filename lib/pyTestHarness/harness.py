import os,sys
import argparse
import pyTestHarness.unittest as unittest
import pyTestHarness.launch as launch
from   pyTestHarness.colors import pthNamedColors as bcolors


class pthHarness:
  def __init__(self,registeredTests):
    self.testsRegistered = 0
    self.testsExecuted = 0
    self.testsSkipped = 0
    self.testsPassed = 0
    self.testsFailed = 0
    self.verbosity_level = 1
    self.pthErrorReportFileName = 'pthErrorReport.log'

    self.testDescription = []
    self.registeredTests = registeredTests

    for t in self.registeredTests:
      if not isinstance(t,unittest.pthUnitTest):
        raise ValueError('[pth]: Registered tests must be of type UnitTest')
    self.testsRegistered = len(self.registeredTests)

    parser = argparse.ArgumentParser(description='Python Test Harness.')
    parser.add_argument('-e', '--execute', help='Execute all tests', required=False, action='store_true')
    parser.add_argument('-v', '--verify', help='Perform test verification only (and not execution)', required=False, action='store_true')
    parser.add_argument('-c', '--configure', help='Configure queuing system information', required=False, action='store_true')
    parser.add_argument('-t', '--test', help='List of test names', required=False)
    parser.add_argument('-o', '--output_path', help='Directory to write stdout into', required=False)
    parser.add_argument('-p', '--purge_output', help='Delete generated output', required=False, action='store_true')
    parser.add_argument('-f', '--error_on_test_failure', help='Return exit code of 1 if any test failed', required=False, action='store_true')
    parser.add_argument('-d', '--configure_default', help='Write default queuing system config file (no mpi, no queuing system)', required=False, action='store_true')
    parser.add_argument('-s', '--sandbox', help='Execute tests in separate directories. Will not work unless you supply absolute paths to executables.', required=False, action='store_true')
    parser.add_argument('-l', '--list', help='List all registered tests and exit', required=False, action='store_true')
    self.args, self.unknown = parser.parse_known_args()

    # If "list" option supplied, simply print out all the tests' names and exit
    if self.args.list :
      for test in registeredTests :
        print(test.name)
      sys.exit(0)

    # If --configure_default is specified, write the default file and exit
    if self.args.configure_default:
      launch.pthLaunch.writeDefaultDefinition()
      sys.exit(0)

    # Create the launcher
    self.launcher = launch.pthLaunch()

    # If --configure is specified, (re)generate the configuration file and exit
    if self.args.configure:
      self.launcher.configure()
      sys.exit(0)

    # Instruct tests to use sandboxes if requested
    if self.args.sandbox :
      for test in registeredTests :
        test.use_sandbox = True

    # Set output path on all tests if --output_path option was included
    if self.args.output_path:
      for test in self.registeredTests:
        test.setOutputPath(self.args.output_path)

    # Label tests as Registered or Excluded:Reason
    for i in range(0,len(self.registeredTests)):
      self.testDescription.append('Registered')

    # Exclude parallel tests if mpiLauncher = 'none' and test uses more than 1 rank
    counter = 0
    skipCounter = 0
    for t in self.registeredTests:
      if self.launcher.mpiLaunch == 'none' and t.ranks != 1:
        self.registeredTests[counter].ignore = True
        self.testDescription[counter] = 'No MPI launcher was provided and test requested > 1 MPI rank'
        skipCounter = skipCounter + 1
      counter = counter + 1

    # Exclude tests based on command line option
    subList = []
    if self.args.test:
      tnames = self.args.test.split(',')
      for name in tnames:
        found = False
        for t in self.registeredTests:
          if name == t.name:
            subList.append(t)
            found = True
            break

        if found == False:
          errstr= '[pth] You requested to test a subset of registered tests, \n\t\t  but no registered test matched the name \"' + name + '\"'
          raise RuntimeError(errstr)

    if self.args.test:
      counter = 0
      for t in self.registeredTests:
        # skip tests already marked to be ignored (e.g. due to mpiLaunch = none and test.ranks != 1
        if t.ignore == True:
          counter = counter + 1
          continue

        if t not in subList:
          t.ignore = True
          self.testDescription[counter] = 'Excluded based on users command line arg -t'
        counter = counter + 1

    # Clean all output unless we are verifying (only)
    if not self.args.verify :
      self.clean()

    # If -p was supplied, stop after cleaning
    if self.args.verify and self.args.purge_output :
      raise RuntimeError('-p (--purge_output) and -v (--verify) are not supported together')
    if self.args.purge_output :
      exit(0);

  def setVerbosityLevel(self,value):
    self.verbosity_level = value
    self.launcher.setVerbosityLevel(self.verbosity_level)

  def execute(self):
    '''Execute tests unless verifying (only) or purging output (only)'''
    if not self.args.verify and not self.args.purge_output:
      print('')
      print(bcolors.HEADER + '[ *** Executing Tests *** ]' + bcolors.ENDC)
      launcher = self.launcher
      testList = self.registeredTests
      description = self.testDescription
      if launcher.verbosity_level > 0:
        launcher.view()

      skipCounter = 0
      for t in testList:
        if launcher.mpiLaunch == 'none' and t.ranks != 1:
          if t.ignore == True:
            skipCounter = skipCounter + 1

      if skipCounter != 0:
        print('\n' + bcolors.WARNING + 'Warning: ' + str(skipCounter) + ' MPI parallel jobs are being skipped as a valid MPI launcher was not provided'+ bcolors.ENDC)

      counter = 0
      for test in testList:
        if test.ignore == False:
          print('[-- Executing test: ' + test.name + ' --]')
          launcher.submitJob(test)
        else:
          print('[-- Skipping test: ' + test.name + ' --] Reason: ' + description[counter])
        counter = counter + 1

  def verify(self):
    '''Verify, unless we are running with a batch system and are not in verify(-only) mode'''
    if not self.launcher.useBatch or self.args.verify :
      print('')
      print(bcolors.HEADER + '[ *** Verifying Test Output *** ]' + bcolors.ENDC)
      tests_not_skipped = 0
      for test in self.registeredTests:
        print('[-- Verifying test: ' + test.name + ' --]')
        if test.ignore :
          print('[Skipping verification for test \"' + test.name + '\"]')
        elif self.launcher.mpiLaunch == 'none' and test.ranks != 1:
          print('[Skipping verification for test \"' + test.name + '\" as test uses > 1 MPI ranks and no MPI launcher was provided]')
        else:
          tests_not_skipped = tests_not_skipped + 1
          test.verifyOutput()

      print('')
      counter = 0
      for test in self.registeredTests:
        if not test.ignore and not test.passed :
          counter = counter + 1
      if counter > 0:
        print('')
        print('[--------- Unit test error report ----------------------]')
        for test in self.registeredTests:
          test.report('log')

      errfile = self.reportAll()
      if errfile and self.args.error_on_test_failure:
        print('\n')
        print('Contents of "' + errfile +'"')
        os.system('cat ' + errfile)
        sys.exit(1)

  def clean(self):
    print('')
    print(bcolors.HEADER + '[ *** Deleting Existing Test Output *** ]' + bcolors.ENDC)
    if os.path.isfile(self.pthErrorReportFileName) :
      os.remove(self.pthErrorReportFileName)
    for test in self.registeredTests:
      if test.ignore:
        print('[ -- Skipping test:',test.name,'--]')
      else:
        self.launcher.clean(test)

  def reportAll(self):
    launcher = self.launcher
    testList = self.registeredTests
    print('')
    nTests = len(testList)
    failCounter = 0
    execCounter = 0
    skipCounter = 0
    seqCounter = 0
    mpiCounter  = 0
    seqPassedCounter = 0
    mpiPassedCounter  = 0
    seqExecCounter = 0
    mpiExecCounter  = 0

    for test in testList:
      if test.ranks == 1:
        seqCounter = seqCounter + 1
      else:
        mpiCounter = mpiCounter + 1

      if test.ignore == False:
        execCounter = execCounter + 1

        if test.ranks == 1:
          seqExecCounter = seqExecCounter + 1
        else:
          mpiExecCounter = mpiExecCounter + 1

        if test.passed == False:
          failCounter = failCounter + 1
      else:
        skipCounter = skipCounter + 1

    for test in testList:
      if test.ignore == False:
        if test.ranks == 1 and test.passed == True:
          seqPassedCounter = seqPassedCounter + 1
        elif test.ranks >= 1 and test.passed == True:
          mpiPassedCounter = mpiPassedCounter + 1

    print('[--------- UnitTest status ----------------------]')
    for test in testList:
      test.report('summary')

    print('')
    print('[--------- UnitTest report ----------------------]')
    print('  ' + ("%.4d" % nTests) + ' ' + 'UnitTests registered')
    print('  ' + ("%.4d" % seqCounter) + ' Sequential UnitTests')
    print('  ' + ("%.4d" % mpiCounter) + ' MPI UnitTests')
    print('  ' + ("%.4d" % execCounter) + ' of ' +("%.4d" % nTests)+ ' UnitTests executed')

    print('')
    if execCounter == 0:
      print(bcolors.WARNING + ' [status] UNKNOWN: All tests were skipped' + bcolors.ENDC)

    elif failCounter > 0:
      print(bcolors.FAIL + ' [status] FAIL: ' + str(failCounter) + ' of ' + str(execCounter) + bcolors.FAIL + ' tests executed FAILED' + bcolors.ENDC)

      print(bcolors.FAIL +'          ' + ("%.4d" % (seqExecCounter-seqPassedCounter)) + ' of ' +("%.4d" % seqExecCounter)+ ' executed Sequential tests failed'+bcolors.ENDC)
      print(bcolors.FAIL +'          ' + ("%.4d" % (mpiExecCounter-mpiPassedCounter)) + ' of ' +("%.4d" % mpiExecCounter)+ ' executed MPI tests failed'+bcolors.ENDC)
    else:
      if skipCounter == 0:
        print(bcolors.OKGREEN + ' [status] SUCCESS: All registered tests passed' + bcolors.ENDC)
      else:
        print(bcolors.WARNING + ' [status] SUCCESS (partial): All executed tests passed' + bcolors.ENDC)

    if seqExecCounter + mpiExecCounter != nTests:
      print(bcolors.WARNING+'          Warning: Not all UnitTests were executed!'+ bcolors.ENDC)
    if seqExecCounter != seqCounter:
      print(bcolors.WARNING+'          Warning: '+("%.4d" % (seqCounter-seqExecCounter))+' sequential UnitTests were skipped'+ bcolors.ENDC)
    if mpiExecCounter != mpiCounter:
      print(bcolors.WARNING+'          Warning: '+("%.4d" % (mpiCounter-mpiExecCounter))+' MPI UnitTests were skipped!'+ bcolors.ENDC)

    errfile = []
    if failCounter > 0:
      file = open(self.pthErrorReportFileName,'w')
      sys.stdout = file

      for test in testList:
        test.report('log')

      file.close()
      sys.stdout = sys.__stdout__

      print('xxx============================================================================xxx')
      print('     UnitTests failed - Full error report written to ' + self.pthErrorReportFileName)
      print('                      - Inspect the error log file and resolve failed tests')
      pthErrorReportFileLocation = os.path.realpath(self.pthErrorReportFileName)
      print('     cat ' + pthErrorReportFileLocation)
      print('xxx============================================================================xxx')
      errfile = pthErrorReportFileLocation
    return(errfile)
