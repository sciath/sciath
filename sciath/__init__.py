""" SciATH: Scientific Application Test Harness """
from sciath._sciath_io import NamedColors

__version__ = (0, 7, 0)

# A default set of colors
SCIATH_COLORS = NamedColors()
