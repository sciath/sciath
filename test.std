# this is a test file

$verify_str abcdf__efsss

  $verify = { 2.2 }


$verify { 
3.33,
6.6 }

KSP its 4

$verify = 99.9

$verify 51.0
