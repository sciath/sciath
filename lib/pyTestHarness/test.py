from __future__ import print_function
import os
import numpy as np
import math as math
import re
from pyTestHarness.colors import NamedColors as pthcolors

def compareLiteral(input,expected):
  status = True
  err = ''
  if len(input) != len(expected):
    status = False
    err = err + "compareLiteral [failed]\nReason: input and expected are of different length\n"
    err = err + ("  expected: %s\n" % expected)
    err = err + ("  input:    %s\n" % input)
    return status,err
  for index in range(0,len(expected)):
    if input[index] != expected[index]:
      status = False
      err = err + "compareLiteral [failed]\nReason: strings are different\n"
      err = err + ("  expected: %s\n" % expected)
      err = err + ("  input:    %s\n" % input)
      err = err + "  index[" + str(index) +  "]" + " input \"" +  input[index] + "\" != expected \"" + expected[index] + "\"\n"
  return status,err

def compareFloatingPointAbsolute(input,tolerance,expected):
  status = True
  err = ''
  tmp = np.array(input)
  i_f = tmp.astype(np.float)
  tmp = np.array(expected)
  e_f = tmp.astype(np.float)
  tol_f = float(tolerance)
  if len(input) != len(expected):
    status = False
    err = err + "compareFloatingPointAbsolute [failed]\nReason: input and expected are of different length\n"
    err = err + ("  expected: %s\n" % e_f)
    err = err + ("  input:    %s\n" % i_f)
    return status,err
  for index in range(0,len(e_f)):
    absdiff = np.abs(i_f[index] - e_f[index]);
    if absdiff > tol_f:
      status = False
      err = err + "compareFloatingPointAbsolute [failed]\nReason: absolute tolerance " + ("%1.4e" % tol_f) + " not satisfied\n"
      err = err + ("  expected: %s\n" % e_f)
      err = err + ("  input:    %s\n" % i_f)
      err = err + "  index[" + str(index) + "]" + (" input \"%1.6e\"" %  i_f[index])  + (" != expected \"%1.6e\"" % e_f[index]) + " (+/-" + ("%1.4e" % tol_f)+" abs.)\n"
  return status,err

# TODO: reduce horrible code duplication here (and in many places in this file)
def compareFloatingPointRelative(input,tolerance,expected):
  status = True
  err = ''
  tmp = np.array(input)
  i_f = tmp.astype(np.float)
  tmp = np.array(expected)
  e_f = tmp.astype(np.float)
  tol_f = float(tolerance)
  if len(input) != len(expected):
    status = False
    err = err + "compareFloatingPointRelative [failed]\nReason: input and expected are of different length\n"
    err = err + ("  expected: %s\n" % e_f)
    err = err + ("  input:    %s\n" % i_f)
    return status,err
  for index in range(0,len(e_f)):
    if e_f[index] == 0.0 :
      reldiff = 0.0 # for exactly zero values, set diff to 0 (controversial)
    else :
      reldiff = np.abs(i_f[index] - e_f[index])/np.abs(e_f[index]);
    if reldiff > tol_f:
      status = False
      err = err + "compareFloatingPointRelative [failed]\nReason: relative tolerance " + ("%1.4e" % tol_f) + " not satisfied\n"
      err = err + ("  expected: %s\n" % e_f)
      err = err + ("  input:    %s\n" % i_f)
      err = err + "  index[" + str(index) + "]" + (" input \"%1.6e\"" %  i_f[index])  + (" != expected \"%1.6e\"" % e_f[index]) + " (+/-" + ("%1.4e" % tol_f)+" rel. )\n"
  return status,err

def compareInteger(input,tolerance,expected):
  status = True
  err = ''
  tmp = np.array(input)
  i_i = tmp.astype(np.int)
  tmp = np.array(expected)
  e_i = tmp.astype(np.int)
  tol_i = int(tolerance)
  if len(input) != len(expected):
    status = False
    err = err + "compareInteger [failed]\nReason: input and expected are of different length\n"
    err = err + ("  expected: %s\n" % e_i)
    err = err + ("  input:    %s\n" % i_i)
    return status,err
  for index in range(0,len(e_i)):
    absdiff = np.abs(i_i[index] - e_i[index]);
    if absdiff > tol_i:
      status = False
      err = err + "compareInteger [failed]\nReason: tolerance " + str(tol_i) + " not satisifed\n"
      err = err + ("  expected: %s\n" % e_i)
      err = err + ("  input:    %s\n" % i_i)
      err = err + "  index[" + str(index) + "]" + (" input \"%s\"" %  i_i[index]) + (" != expected \"%s\"" % e_i[index]) + " (+/-"+str(tol_i)+")\n"
  return status,err

def parseFile(filename,keywords):
  flat = ''
  contents = []
  if filename:
    file = open(filename,"r")
    for line in file:
      if line.rstrip():
        if not any(keywords in line for keywords in keywords):
          rm_lb = line.lstrip()
          rm_lb = rm_lb.rstrip()
          flat = flat + (rm_lb + ' ')
          stripped_line = line.rstrip()
          stripped_line = stripped_line.lstrip()
          contents.append(stripped_line)
    file.close()
  return(contents,flat)

def getKeyValues(contents,keyword):
  c1 = contents
  f_c = []
  f_r = []
  f_s = []

  # Look for arrays, then strip them out
  find_curly  = re.findall("\s*" + keyword + "\s*=?\s*[\{].*?[\}]", c1)
  if find_curly:
    f_c  = re.findall("\s*" + keyword + "\s*=?\s*[\{](.*?)[\}]", c1)
    for item in find_curly:
      c1 = c1.replace(item,' ')
  find_round  = re.findall("\s*" + keyword + "\s*=?\s*[\(].*?[\)]", c1)
  if find_round:
    f_r  = re.findall("\s*" + keyword + "\s*=?\s*[\(](.*?)[\)]", c1)
    for item in find_round:
      c1 = c1.replace(item,' ')
  find_square  = re.findall("\s*" + keyword + "\s*=?\s*[\[].*?[\]]", c1)
  if find_square:
    f_s  = re.findall("\s*" + keyword + "\s*=?\s*[\[](.*?)[\]]", c1)
    for item in find_square:
      c1 = c1.replace(item,' ')
  find_single = []
  f_single = re.findall("\s*" + keyword + "\s*=?\s*(.*?)\s", c1)
  list = f_c + f_r + f_s + f_single
  filtered = []
  if list:
    for item in list:
      if item != '':
        trimmed = item.lstrip()
        trimmed = trimmed.rstrip()
        filtered.append(trimmed)
  return filtered

def getKeyValuesAsInt(contents,keyword):
  result = getKeyValues(contents,keyword)
  flattened = []
  for sublist in result:
    if type(sublist) is list:
        pass
    else:
      if sublist != '':
        numlist = sublist.replace(',',' ')
        numlist = numlist.replace(';',' ')
        nums = numlist.split(' ')
        #print(nums)
        for num in nums:
          if num != '':
            flattened.append(num)
  tmp = np.array(flattened)
  values = tmp.astype(np.int)
  return values

def getKeyValuesAsFloat(contents,keyword):
  result = getKeyValues(contents,keyword)
  flattened = []
  for r in result:
    r = r.replace(',',' ')
    r = r.replace(';',' ')
    val = r.split(' ')
    for v in val:
      if v != '':
        flattened.append(v)
  tmp = np.array(flattened)
  values = tmp.astype(np.float)
  return values

def getKeyValuesNLinesInclusive(contents,keyword,numlines):
  result = []
  counter = 0
  start = 0
  for line in contents:
    x = line.find(keyword)
    if x != -1:
      start = counter
      break
    counter = counter + 1
  for index in range(start,start+numlines):
    result.append(contents[index])
  return result

def getKeyValuesNLinesExclusive(contents,keyword,numlines):
  result = []
  counter = 0
  start = 0
  for line in contents:
    x = line.find(keyword)
    if x != -1:
      start = counter + 1
      break
    counter = counter + 1
  for index in range(start,start+numlines):
    result.append(contents[index])
  return result


class Test:
  def __init__(self,name,ranks,execute,expected_file):
    self.passed = -1
    self.walltime = 2.0 # minutes
    self.errormessage = ''
    self.errno = -1
    self.name = name
    if not name or len(name) == 0 :
      raise RuntimeError('Tests must be named')
    self.ranks = ranks
    if isinstance(execute,list):
      self.execute = execute
    if isinstance(execute,str):
      self.execute = [execute]
    self.expected_file = expected_file
    self.keywords = [ '#', '!', '//' ]
    self.output_file = name + '-p' + str(ranks) + '.output'
    self.comparison_file = ''
    self.output_path = ''
    self.sandbox_path = self.name + '_sandbox'
    self.use_sandbox = False
    self.ignore = False
    self.verbosity_level = 1

  def setVerbosityLevel(self,value):
    self.verbosity_level = value

  def verify(self,junk):
    errstr = '[pth] A valid verification method for test \"' + self.name + '\" was not found.\n\
              [pth] You must provide each test with a method to verify the output.\n\
              [pth] The method is set via calling test.setVerifyMethod()'
    raise RuntimeError(errstr)

  def setOutputPath(self,opath):
    self.output_path = opath

  def setComparisonFile(self,fname):
    self.comparison_file = fname

  def setVerifyMethod(self,verify):
    self.verify = verify

  def setWalltime(self,mins):
    self.walltime = mins

  def setUseSandbox(self) :
    self.use_sandbox=True

  def appendKeywords(self,keywords):
    self.keywords.append(keywords)
  #print(self.keywords)

  def getExitCode(self):
    return self.errno

  def getErrorReport(self):
    return self.errormessage

  def getErrorStatus(self):
    return self.passed

  def verifyOutput(self):
    if self.use_sandbox :
      sandboxBack = os.getcwd()
      os.chdir(self.sandbox_path)
    if self.comparison_file == '':
      self.comparison_file = os.path.join(self.output_path,self.output_file)
    if self.verbosity_level > 0:
      print('[Parsing file]',self.comparison_file)
    (self.output_contents,self.output_flatcontents) = parseFile(self.comparison_file,self.keywords)
    if self.use_sandbox :
      os.chdir(sandboxBack)
    if self.verbosity_level > 0:
      print('[Parsing file]',self.expected_file)
    (self.expected_contents,self.expected_flatcontents) = parseFile(self.expected_file,self.keywords)
    self.verify(self)

  def getOutput(self):
    return self.output_contents,self.output_flatcontents

  def getExpected(self):
    return self.expected_contents,self.expected_flatcontents

  def updateStatus(self,status,err):
    if self.passed == -1:
      self.passed = status
      if err != '':
        self.errormessage = self.errormessage + err
      return
    if status == False:
      self.passed = False
      if err != '':
        self.errormessage = self.errormessage + err

  def report(self,type):
    if type == 'summary':
      if self.ignore == True:
        print(pthcolors.WARNING + ' [' + self.name + ']   skipped' + pthcolors.ENDC)
      else:
        if self.passed == False:
          print(pthcolors.FAIL + ' [' + self.name + ']   *** FAILED ***' + pthcolors.ENDC)
        else:
          print(pthcolors.OKGREEN + ' [' + self.name + ']   passed' + pthcolors.ENDC)
    if type == 'log':
      if self.ignore == False:
        if self.passed == False:
          print('[' + self.name + '] reason for failure\n' + '--------------------------------------------------------------\n' + self.errormessage)

  def compareFloatingPointAbsolute(self,key,tolerance):
    expected,expected_flat = self.getExpected()
    output,output_flat = self.getOutput()
    values_e = getKeyValuesAsFloat(expected_flat,key)
    if len(values_e) == 0:
      errstr = '[pth][VerificationError] Test \"' + self.name + '\" queried the expected file \"' + self.expected_file + '\" for key \"' + key + '\" which was not found. \n\t\t    Users verification code is likely incorrect (contains a typo in the key name)'
      raise RuntimeError(errstr)
    values   = getKeyValuesAsFloat(output_flat,key)
    status,err = compareFloatingPointAbsolute(values,tolerance,values_e)
    kerr = ''
    if status == False:
      kerr = 'Key = \"' + key + '\" --> ' + err
    self.updateStatus(status,kerr)

  #Deprecated : default to absolute test
  compareFloatingPoint=compareFloatingPointAbsolute

  def compareFloatingPointRelative(self,key,tolerance):
    expected,expected_flat = self.getExpected()
    output,output_flat = self.getOutput()
    values_e = getKeyValuesAsFloat(expected_flat,key)
    if len(values_e) == 0:
      errstr = '[pth][VerificationError] Test \"' + self.name + '\" queried the expected file \"' + self.expected_file + '\" for key \"' + key + '\" which was not found. \n\t\t    Users verification code is likely incorrect (contains a typo in the key name)'
      raise RuntimeError(errstr)
    values   = getKeyValuesAsFloat(output_flat,key)
    status,err = compareFloatingPointRelative(values,tolerance,values_e)
    kerr = ''
    if status == False:
      kerr = 'Key = \"' + key + '\" --> ' + err
    self.updateStatus(status,kerr)

  def compareInteger(self,key,tolerance):
    expected,expected_flat = self.getExpected()
    output,output_flat = self.getOutput()
    values_e = getKeyValuesAsInt(expected_flat,key)
    if len(values_e) == 0:
      errstr = '[pth][VerificationError] Test \"' + self.name + '\" queried the expected file \"' + self.expected_file + '\" for key \"' + key + '\" which was not found. \n\t\t    Users verification code is likely incorrect (contains a typo in the key name)'
      raise RuntimeError(errstr)

    values   = getKeyValuesAsInt(output_flat,key)
    status,err = compareInteger(values,tolerance,values_e)
    kerr = ''
    if status == False:
      kerr = 'Key = \"' + key + '\" --> ' + err
    self.updateStatus(status,kerr)

  def compareLiteral(self,key):
    expected,expected_flat = self.getExpected()
    output,output_flat = self.getOutput()
    values_e = getKeyValues(expected_flat,key)
    if len(values_e) == 0:
      errstr = '[pth][VerificationError] Test \"' + self.name + '\" queried the expected file \"' + self.expected_file + '\" for key \"' + key + '\" which was not found. \n\t\t    Users verification code is likely incorrect (contains a typo in the key name)'
      raise RuntimeError(errstr)
    values   = getKeyValues(output_flat,key)
    status,err = compareLiteral(values,values_e)
    kerr = ''
    if status == False:
      kerr = 'Key = \"' + key + '\" --> ' + err
    self.updateStatus(status,kerr)

  def compareUnixDiff(self):
    expected,expected_flat = self.getExpected()
    output,output_flat = self.getOutput()
    status,err = compareLiteral(output,expected)
    kerr = ''
    if status == False:
      kerr = 'Key = \"' + '<entire file>' + '\" --> ' + err
    self.updateStatus(status,kerr)

  def clean(self):
    outfile = os.path.join(self.output_path,self.output_file)
    cmpfile = self.comparison_file
    print('<test> rm -f ' + outfile)
    if outfile != cmpfile and cmpfile:
      print('<test> rm -f ' + cmpfile)
