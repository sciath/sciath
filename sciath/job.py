""" SciATH Job class """
from __future__ import print_function

import sciath.task


class Job:
    r""":class:`Job` describes a ordered list of :class:`Task`\s.

    It describes the :class:`Task`\s themselves, not information about any
    particular "run" (handled within :class:`Harness`) or how one might
    interpret the results of such a run (handled by :class:`Test`).  A
    :class:`Launcher` object executes the :class:`Task`\s described by a :class:`Job`.

    Data include

    * A name
    * An ordered list of :class:`Task`\s

    It also defines how the name may be converted to various standard filenames.

    """

    _default_job_name = 'job'

    def __init__(self, task_or_tasks, name=None):
        if name is None:
            self.name = Job._default_job_name
            self.named_by_default = True
        else:
            self.name = name.replace(' ', '_')
            self.named_by_default = False

        if isinstance(task_or_tasks, sciath.task.Task):
            self.tasks = [task_or_tasks]
        elif isinstance(task_or_tasks, list):
            for task in task_or_tasks:
                if not isinstance(task, sciath.task.Task):
                    raise Exception('A Task or list of Tasks was expected')
                self.tasks = task_or_tasks
        else:
            raise Exception('A Task or list of Tasks was expected')

    def create_execute_command(self):
        """
        Returns a list containing (command, resource) tuples for the Job.
        """
        return [task.create_execute_command() for task in self.tasks]

    @property
    def complete_filename(self):
        """ Returns a filename for a sentinel signifying completion """
        return '.%s.complete' % self.name

    @property
    def exitcode_filename(self):
        """ Returns a filename to use for exit codes """
        return self.name + '.exitcode'

    @property
    def launched_filename(self):
        """ Returns a filename for a sentinel signifying launching """
        return '.%s.launched' % self.name

    @property
    def stdout_filename(self):
        """ Returns a filename to use for stdout """
        return self.name + '.stdout'

    @property
    def stderr_filename(self):
        """ Returns a filename to use for stderr """
        return self.name + '.stderr'

    # This can be removed once the transition to using templates is complete
    def get_max_resources(self):
        """ Returns a dict() defining the maximum required counts / values """

        # Use the first Task to determine which resources to consider
        max_resources = {}
        for key, value in self.tasks[0].resources.items():
            max_resources[key] = value

        for task in self.tasks[1:]:
            for key, value in task.resources.items():
                if value > max_resources[key]:
                    max_resources[key] = value
        return max_resources

    def resource_max(self, key):
        """ Returns the maximum value of a resource over all Tasks

            Returns None if no Tasks define that resource.
        """
        maximum = None
        for task in self.tasks:
            if key in task.resources:
                current = task.resources[key]
                if maximum is None:
                    maximum = current
                else:
                    maximum = max(maximum, current)
        return maximum

    def number_tasks(self):
        """ Returns the number of tasks within the Job """
        return len(self.tasks)

    def total_wall_time(self):
        """ Returns the total wall time required for all Tasks

            Returns None if any task does not have a defined wall time.
        """
        total_wall_time = 0
        for task in self.tasks:
            if task.wall_time is None:
                return None
            total_wall_time += task.wall_time
        return total_wall_time
