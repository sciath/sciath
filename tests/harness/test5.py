#!/usr/bin/env python

# A test using a LineVerifier

import os

import sciath.harness
import sciath.test
import sciath.job
import sciath.verifier_line


this_dir = os.path.dirname(os.path.realpath(__file__))

command = ['echo','The first number is 1.1\nThe second number is 1.01']
job = sciath.job.Job(command)
test = sciath.test.Test(job,'test')
verifier = sciath.verifier_line.LineVerifier(test, expected_file = os.path.join(this_dir,'test5_data','test.expected'))
verifier.rules.append(sciath.verifier_line.key_and_float_rule('The first number is'))
verifier.rules.append(sciath.verifier_line.key_and_float_rule('The second number is'))
test.verifier = verifier

harness = sciath.harness.Harness([test])
harness.run_from_args()
