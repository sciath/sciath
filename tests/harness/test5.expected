[ -- Removing output for Test: test -- ]
[36m[Executing job][0m from <<PATH STRIPPED>>
['echo', 'The first number is 1.1\nThe second number is 1.01']
--- Error Reports ---
[Report for test]
Report for lines matching: '^The\ first\ number\ is'
(l.1/1) Lines did not match:
expected value of 7 does not match output value 1.1 to rel. tol. 1e-06
-- Summary --
test : fail (verification failed)
TEST SUCCESS NOT CONFIRMED
